### PR TITLE
energy-management-app: refactor WH and EVSE into separate endpoints and fix conformance issues

### PR DESCRIPTION
**Pull request by soares-sergio**
This PR refactors the energy-management-app into 2 separate endpoints
(EP1 for EVSE and EP2 for WaterHeater) and fixes almost all of the device conformance issues.

Main contributions from this PR:
* `TC_DeviceBasicComposition.py` failed on this app before this PR and now passes.
* `TC_DeviceConformance.py` failed more than 13 tests for each app and now only fails one for `evse` and two for water-`heater`.

Changes:
* Split Water Heater and EVSE into two separate endpoints (1 and 2).
* Dynamically disable unused endpoint at runtime. Based on the app choice
  (command line argument on linux [`evse` of `water-heater`] or #define in ESP32 or SIlabs). 
  * For example, for Water Heater, initialize clusters on endpoint 2 and
  disable endpoint 1 (EVSE).
* Refactor/move the cluster init code related to ElectricalSensor (PowerTopology, EPM and EEM) and DEM out of `evse` and into a common directory that can be shared between EVSE and WaterHeater.
* Fix many conformance issues:
  * Bump outdated cluster revisions
  * Fix feature map bits
  * Fix device description for each endpoint
  * Remove extra clusters
  * Move PowerSource from EP0 to EP1
  * When using `evse`, only unitialize `evse` related clusters. Same for WaterHeater  
* Update `TC_WHM_xyz` test scripts to use the correct endpoint passed in as argument

Tests performed:
1. Check basic composition for EVSE:
```
scripts/run_in_python_env.sh out/python_env './scripts/tests/run_python_test.py --app ./out/linux-x64-energy-management-no-ble/chip-energy-management-app --app-args "--application evse --trace-to json:log" --script src/python_testing/TC_DeviceBasicComposition.py --script-args "--qr-code MT:-24J0AFN00KA0648G00"'
```

2. Check basic composition for WaterHeater:
```
scripts/run_in_python_env.sh out/python_env './scripts/tests/run_python_test.py --app ./out/linux-x64-energy-management-no-ble/chip-energy-management-app --app-args "--application water-heater --trace-to json:log" --script src/python_testing/TC_DeviceBasicComposition.py --script-args "--qr-code MT:-24J0AFN00KA0648G00"'
```

3. Check evse app against `TC_EEVSE_2_6.py`:
```
rm -f evse.bin; ./out/linux-x64-energy-management-no-ble/chip-energy-management-app --enable-key 000102030405060708090a0b0c0d0e0f --KVS evse.bin --featureSet 0x3d

python src/python_testing/TC_EEVSE_2_6.py --endpoint 1 -m on-network -n 1234 -p 20202021 -d 3840 --hex-arg enableKey:000102030405060708090a0b0c0d0e0f
```
4. Verified it builds for esp32, silabs and linux.
5. Verified all-clusters-app builds
6. Verified `TC_WHM_1_2` and `TC_WHM_2_1`

**EVSE app Conformance failures before this PR:**
```
chip-energy-management-app,"  Problem: ProblemSeverity.ERROR
  test_name: IDM-10.3
  location:
  Endpoint: 0,
  Cluster:  40 (0x28) BasicInformation
  Attribute:65533 (0xfffd)
  problem: Revision found on cluster (3) does not match revision listed in the spec (4)"
chip-energy-management-app,"  Problem: ProblemSeverity.ERROR
  test_name: IDM-10.3
  location:
  Endpoint: 0,
  Cluster:  47 (0x2f) PowerSource
  Attribute:65533 (0xfffd)
  problem: Revision found on cluster (2) does not match revision listed in the spec (3)"
chip-energy-management-app,"  Problem: ProblemSeverity.ERROR
  test_name: IDM-10.3
  location:
  Endpoint: 0,
  Cluster:  48 (0x30) GeneralCommissioning
  Attribute:65533 (0xfffd)
  problem: Revision found on cluster (1) does not match revision listed in the spec (2)"
chip-energy-management-app,"  Problem: ProblemSeverity.ERROR
  test_name: IDM-10.3
  location:
  Endpoint: 1,
  Cluster:  3 (0x03) Identify
  Attribute:65533 (0xfffd)
  problem: Revision found on cluster (4) does not match revision listed in the spec (5)"
chip-energy-management-app,"  Problem: ProblemSeverity.ERROR
  test_name: IDM-10.3
  location:
  Endpoint: 1,
  Cluster:  148 (0x94) WaterHeaterManagement
  Attribute:65533 (0xfffd)
  problem: Revision found on cluster (ValueDecodeFailure(TLVValue=None, Reason=InteractionModelError(<Status.Failure: 1>))) does not match revision listed in the spec (2)"
chip-energy-management-app,"  Problem: ProblemSeverity.ERROR
  test_name: IDM-10.3
  location:
  Endpoint: 1,
  Cluster:  153 (0x99) EnergyEvse
  Attribute:65533 (0xfffd)
  problem: Revision found on cluster (2) does not match revision listed in the spec (3)"
chip-energy-management-app,"  Problem: ProblemSeverity.ERROR
  test_name: IDM-10.3
  location:
  Endpoint: 1,
  Cluster:  157 (0x9d) EnergyEvseMode
  Attribute:65533 (0xfffd)
  problem: Revision found on cluster (1) does not match revision listed in the spec (2)"
chip-energy-management-app,"  Problem: ProblemSeverity.ERROR
  test_name: IDM-10.3
  location:
  Endpoint: 1,
  Cluster:  159 (0x9f) DeviceEnergyManagementMode
  Attribute:65533 (0xfffd)
  problem: Revision found on cluster (1) does not match revision listed in the spec (2)"
chip-energy-management-app,"  Problem: ProblemSeverity.ERROR
  test_name: IDM-10.5
  location:
  Endpoint: 1,
  Cluster:  152 (0x98) DeviceEnergyManagement
  problem: Extra cluster found on endpoint with device types [Descriptor.Structs.DeviceTypeStruct(deviceType=1292, revision=1), Descriptor.Structs.DeviceTypeStruct(deviceType=1296, revision=1)]
  spec_location:"
chip-energy-management-app,"  Problem: ProblemSeverity.ERROR
  test_name: IDM-10.5
  location:
  Endpoint: 1,
  Cluster:  148 (0x94) WaterHeaterManagement
  problem: Extra cluster found on endpoint with device types [Descriptor.Structs.DeviceTypeStruct(deviceType=1292, revision=1), Descriptor.Structs.DeviceTypeStruct(deviceType=1296, revision=1)]
  spec_location:"
chip-energy-management-app,"  Problem: ProblemSeverity.ERROR
  test_name: IDM-10.5
  location:
  Endpoint: 1,
  Cluster:  158 (0x9e) WaterHeaterMode
  problem: Extra cluster found on endpoint with device types [Descriptor.Structs.DeviceTypeStruct(deviceType=1292, revision=1), Descriptor.Structs.DeviceTypeStruct(deviceType=1296, revision=1)]
  spec_location:"
chip-energy-management-app,"  Problem: ProblemSeverity.ERROR
  test_name: IDM-10.5
  location:
  Endpoint: 1,
  Cluster:  159 (0x9f) DeviceEnergyManagementMode
  problem: Extra cluster found on endpoint with device types [Descriptor.Structs.DeviceTypeStruct(deviceType=1292, revision=1), Descriptor.Structs.DeviceTypeStruct(deviceType=1296, revision=1)]
  spec_location:"
```

EVSE app conformance failures after this PR (only one :)):
```
Problem: ProblemSeverity.ERROR
    test_name: IDM-10.2
    location: 
       Endpoint: 0,
       Cluster:  49 (0x31) NetworkCommissioning
      Attribute:8 (0x08)
    problem: Attribute 0x08 is included, but is disallowed by conformance. Conformance: WI, implemented features: ET
    spec_location: 
```
This failure is expected as the example apps have both ET and WI implemented, while this is disallowed in a real product.

### CODE REVIEW

File: examples/energy-management-app/energy-management-common/common/include/EnergyManagementAppCmdLineOptions.h
Comment (tcarmelveilleux) [inline examples/energy-management-app/energy-management-common/common/include/EnergyManagementAppCmdLineOptions.h:31]: This likely don't belong in a file called `examples/energy-management-app/energy-management-common/common/include/EnergyManagementAppCmdLineOptions.h` when they don't relate to command line options. Suggest moving these elsewhere.

File: examples/energy-management-app/energy-management-common/device-energy-management/src/DEMInit.cpp
Comment (tcarmelveilleux) [inline examples/energy-management-app/energy-management-common/device-energy-management/src/DEMInit.cpp:43]: Anonymous namespace instead of static everywhere please.

File: examples/energy-management-app/energy-management-common/energy-reporting/include/ElectricalSensorInit.h
Comment (tcarmelveilleux) [inline examples/energy-management-app/energy-management-common/energy-reporting/include/ElectricalSensorInit.h:27]: Since these are not widely used it would be good to leave inside the the place that inits everything and injects it down from the places that need it, rather than having to go through these global getters that init more globals in their guts.

Comment (tcarmelveilleux) [reply examples/energy-management-app/energy-management-common/energy-reporting/include/ElectricalSensorInit.h:27]: It may even be worthwhile to introduce a class to hold this state.

**Review by tcarmelveilleux — APPROVED** [review]

File: examples/energy-management-app/energy-management-common/energy-reporting/include/ElectricalSensorInit.h
Comment (soares-sergio) [reply examples/energy-management-app/energy-management-common/energy-reporting/include/ElectricalSensorInit.h:27]: Got rid of those. Now we have a a `CommonMain.h` which initializes the common clusters (DEM, PT, EPM, etc.) and then from there, it initializes either EVSE or WaterHeater and injects the references down, instead of using a global getter. Thanks for the suggestion!

File: examples/energy-management-app/energy-management-common/common/include/EnergyManagementAppCmdLineOptions.h
Comment (soares-sergio) [reply examples/energy-management-app/energy-management-common/common/include/EnergyManagementAppCmdLineOptions.h]: Moved to `CommonMain.h` thanks

File: src/app/zap-templates/zcl/data-model/chip/matter-devices.xml
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/matter-devices.xml:1]: Why all the whitespace-only changes to this file?

Comment (soares-sergio) [reply src/app/zap-templates/zcl/data-model/chip/matter-devices.xml:1]: Good catch, that was a mistake. Fixed, thanks!

**Comment by soares-sergio** [comment]
@jamesharrow @PeterC1965 from the git history, looks like you both are familiar with this app. Would you mind taking a look at this PR? I've only tested these changes on linux. Thanks!

**Comment by lboue** [comment]
@soares-sergio I notice that the values returned are always zero. Is there an option to activate to correct this?
```
    ./chip-tool waterheatermanagement read tank-volume 1 2
    [1730241520.721] [33588:33590] [TOO] Endpoint: 2 Cluster: 0x0000_0094 Attribute 0x0000_0002 DataVersion: 2890564126
    [1730241520.721] [33588:33590] [TOO]   TankVolume: 0

     ./chip-tool waterheatermanagement read tank-percentage 1 2
    [1730241866.744] [33693:33695] [TOO] Endpoint: 2 Cluster: 0x0000_0094 Attribute 0x0000_0004 DataVersion: 2890564126
    [1730241866.744] [33693:33695] [TOO]   TankPercentage: 0

    ./chip-tool waterheatermanagement read boost-state 1 2
    [1730241738.703] [33659:33661] [TOO] Endpoint: 2 Cluster: 0x0000_0094 Attribute 0x0000_0005 DataVersion: 2890564126
    [1730241738.705] [33659:33661] [TOO]   BoostState: 0
```

**Steps (water-heater)**

```
cd ~/connectedhomeip/examples/energy-management-app/linux
gn gen out/debug
ninja -C out/debug -j2
./out/debug/chip-energy-management-app --application water-heater --trace-to json:log
```

**Comment by soares-sergio** [comment]
> @soares-sergio I notice that the values returned are always zero. Is there an option to activate to correct this?
> 
> ```
> ./chip-tool waterheatermanagement read tank-volume 1 2
> [1730241520.721] [33588:33590] [TOO] Endpoint: 2 Cluster: 0x0000_0094 Attribute 0x0000_0002 DataVersion: 2890564126
> [1730241520.721] [33588:33590] [TOO]   TankVolume: 0
> 
> ./chip-tool waterheatermanagement read tank-percentage 1 2
> [1730241866.744] [33693:33695] [TOO] Endpoint: 2 Cluster: 0x0000_0094 Attribute 0x0000_0004 DataVersion: 2890564126
> [1730241866.744] [33693:33695] [TOO]   TankPercentage: 0
> 
> ./chip-tool waterheatermanagement read boost-state 1 2
> [1730241738.703] [33659:33661] [TOO] Endpoint: 2 Cluster: 0x0000_0094 Attribute 0x0000_0005 DataVersion: 2890564126
> [1730241738.705] [33659:33661] [TOO]   BoostState: 0
> ```
> 
> **Steps (water-heater)**
> 
> ```
> cd ~/connectedhomeip/examples/energy-management-app/linux
> gn gen out/debug
> ninja -C out/debug -j2
> ./out/debug/chip-energy-management-app --application water-heater --trace-to json:log
> ```

@lboue : I believe this is expected. I just checked the code on both `ToT` branch and my `PR` and I noticed this example app doesn't configure most of its attribute values by default. But it does work fine if you use `TestEventTriggers`.
For example, the steps below set the default TestEventTrigger which `Simulate installation in a 100L tank full of water at 20C, with a target temperature of 60C, in OFF mode`: https://github.com/project-chip/connectedhomeip/blob/c0071eb67552cae7d12ef47b4f42f57e7b881b20/src/app/clusters/water-heater-management-server/WaterHeaterManagementTestEventTriggerHandler.h#L47

Steps-by-step:
* Build app:
```
./scripts/build/build_examples.py --target linux-x64-energy-management-no-ble build
```

* Run app:
```
rm /tmp/chip_* && ./out/linux-x64-energy-management-no-ble/chip-energy-management-app --application water-heater --trace-to json:log --enable-key 000102030405060708090a0b0c0d0e0f
```

* Commission with chip-tool:
```
./out/linux-x64-chip-tool-no-ble/chip-tool pairing code 0x12344321 MT:-24J0AFN00KA0648G00
```

* Read TankVolume attribute (expect 0 by default):
```
./out/linux-x64-chip-tool-no-ble/chip-tool waterheatermanagement read tank-volume 0x12344321 2 | grep TOO

[1730306361.511] [2089549:2089552] [TOO]   TankVolume: 0
```

* Install default TestTrigger for WaterHeater tank 100L (`0x0094000000000000`):

```
./out/linux-x64-chip-tool-no-ble/chip-tool generaldiagnostics test-event-trigger hex:000102030405060708090a0b0c0d0e0f 0x0094000000000000 0x12344321 0
```

* Read TankVolume attribute (now expect 100):
```
./out/linux-x64-chip-tool-no-ble/chip-tool waterheatermanagement read tank-volume 0x12344321 2 | grep TOO


[1730312762.703] [2153606:2153609] [TOO]   TankVolume: 100
```

**Comment by lboue** [comment]
For Silabs EFF32:
Why not use WaterHeaterApplicationShutdown() here?

https://github.com/project-chip/connectedhomeip/blob/50fb8159c0d9059d9a14f79604fc3903ed4ee35d/examples/energy-management-app/silabs/src/AppTask.cpp#L175

**Comment by soares-sergio** [comment]
> For Silabs EFF32 I am facing an build issue:
> 
> ```
> ./scripts/examples/gn_silabs_example.sh ./examples/energy-management-app/silabs/ ./out/waterheater-app BRD2703A
> (edit)
> ../../../examples/energy-management-app/silabs/src/AppTask.cpp: In function 'void ApplicationShutdown()':
> ../../../examples/energy-management-app/silabs/src/AppTask.cpp:175:5: error: 'FullWhmApplicationShutdown' was not declared in this scope; did you mean 'EvseApplicationShutdown'?
>   175 |     FullWhmApplicationShutdown();
>       |     ^~~~~~~~~~~~~~~~~~~~~~~~~~
>       |     EvseApplicationShutdown
> ```
> 
> `FullWhmApplicationShutdown` is not known in `AppTask.cpp`.
> 
> https://github.com/project-chip/connectedhomeip/blob/f212d601738728d58e88ada3bc7e65b84cdfe7f0/examples/energy-management-app/silabs/src/AppTask.cpp#L160-L161

Thanks @lboue ! Just fixed it.

**Review by jamesharrow — COMMENTED** [review]
General point - 
the Device composition needs to be changed to have:
- EVSE *OR* WaterHeater on EP1 (if we really want both in the same app at the same time then I'll be surprised).
- Electrical Sensor device type should be on EP2 with EEM/EPM
- DEM Utility device type should be on EP3
 - We could have EP4 for Water heater?

At least that is my understanding - it was only because historically the Test Harness makes it very difficult to specific which end point to test on we put all of them on EP1 (which is not allowed).

I was also considering having 2 or more ZAP files in the directory (unless there is an _easy_ way to do this programmatically) - selected at build time to decide which to use (EVSE or Water Heater)... and in the future may add Solar PV and Battery Storage to this list as well.

**Comment by lboue** [comment]
@soares-sergio You should update to last commit because `Tests / REPL Tests CI test `failed.

**Comment by soares-sergio** [comment]
> @soares-sergio You should update to last commit because `Tests / REPL Tests` CI test fails sometime.

Thanks, @lboue !
WaterHeater is now on endpoint 2. I updated the CI test to test look for the correct endpoint on the latest commit. This should be fixed now.

**Comment by soares-sergio** [comment]
> General point - the Device composition needs to be changed to have:
> 
> * EVSE _OR_ WaterHeater on EP1 (if we really want both in the same app at the same time then I'll be surprised).
> * Electrical Sensor device type should be on EP2 with EEM/EPM
> * DEM Utility device type should be on EP3
> * We could have EP4 for Water heater?
> 
> At least that is my understanding - it was only because historically the Test Harness makes it very difficult to specific which end point to test on we put all of them on EP1 (which is not allowed).
> 
> I was also considering having 2 or more ZAP files in the directory (unless there is an _easy_ way to do this programmatically) - selected at build time to decide which to use (EVSE or Water Heater)... and in the future may add Solar PV and Battery Storage to this list as well.

Thanks, @jamesharrow ! These are good points. Maybe we can do this step-by-step? I believe this PR moves things forward in the right direction.

Today on ToT we have:
```
EP0
- EP1 - (EVSE, WaterHeater), EEM, EPM, TP, DEM
```
_Note: One of EVSE or WaterHeater are "half-dead" (not initialized based on a command-line `--application` argument or build time config) but still present when scanned by a client, which causes tests to fail when we try to read attributes from non-initialized clusters._

What I'm proposing with this PR is:
```
EP0
- EP1(EVSE): EEM, EPM, TP, DEM
- EP2(WaterHeater): EEM, EPM, TP, DEM
```
_Note: Only EP1 or EP2 are enabled at a time based on a command-line `--application` argument. The whole endpoint is disabled, i.e., it's not shown when scanned by a client._

You mentioned:
> * Electrical Sensor device type should be on EP2 with EEM/EPM
> * DEM Utility device type should be on EP3

I think that's a good idea. We could still do this in a future PR if we want. In fact, this PR already parameterize some of the clusters' endpointIds, so it will be easier to split them into a different endpoint.

Re:
> I was also considering having 2 or more ZAP files in the directory (unless there is an _easy_ way to do this programmatically) - selected at build time to decide which to use (EVSE or Water Heater)... and in the future may add Solar PV and Battery Storage to this list as well.

I agree multiple zap files is easier for people that want a "grab and go" example. On the other hand, ZAP files go out of date very easily and will add multiple build time configurations. This makes it more complex/slow to test everything. For example, CI would need rebuild this app multiple times, for multiple configurations, etc.
I also noticed the ZAP config goes out of date frequently. For example, a lot of the current example apps have outdated cluster revision. If we add more ZAP files it will increase the maintenance burden on simple updates like this.

Regarding _"not having `EP1` enabled and clients using `EP1` to derive the device type"_, I'm not super familiar with that yet but I discussed it internally and the feedback I got from my teammates was:
 _**Actual commissioners are never confused if they don't find an EP1. Most clients do NOT use the device type of EP1 to determine the "primary" device type, and rather look at the composition of the node and children endpoints**_.

What do you think? Are you OK with moving forward with this PR given what I discussed above?

**Comment by lboue** [comment]
It seems there is an issue with `WaterHeaterMode.Attributes.SupportedModes` attribute:

```
[2024-11-07 17:54:28.881678][TEST][STDERR]INFO:chip.native.FP:Forgetting fabric 0x1
[2024-11-07 17:54:28.882097][TEST][STDOUT]mobly.signals.TestFailure: Details=Error reading <class 'chip.clusters.Objects.WaterHeaterMode'>:<class 'chip.clusters.Objects.WaterHeaterMode.Attributes.SupportedModes'> = ValueDecodeFailure(TLVValue=None, Reason=InteractionModelError(<Status.UnsupportedEndpoint: 127>)), Extras=None
[2024-11-07 17:54:28.882637][TEST][STDOUT][MatterTest] 11-07 17:54:28.873 INFO 		***** Test Failure : 
[2024-11-07 17:54:28.882724][TEST][STDERR]INFO:chip.native.TS:Pending Last Known Good Time: 2023-10-14T01:16:48
[2024-11-07 17:54:28.863686][APP ][STDOUT][25075:25075] CHIP:EM: Rxd Ack; Removing MessageCounter:265997561 from Retrans Table on exchange 3726r
[2024-11-07 17:54:28.883327][TEST][STDOUT][MatterTest] 11-07 17:54:28.873 INFO Finished test in 8564.0ms
[2024-11-07 17:54:28.883629][TEST][STDERR]INFO:chip.native.TS:Previous Last Known Good Time: 2023-10-14T01:16:48
[2024-11-07 17:54:28.864435][APP ][STDOUT][25075:25075] CHIP:ATM: {
[2024-11-07 17:54:28.883748][TEST][STDOUT][MatterTest] 11-07 17:54:28.873 ERROR 
```

**Comment by lboue** [comment]
Could you restart the `Tests / REPL Tests - Linux` CI job ?

**Comment by kiel-apple** [comment]
@lboue re-ran that job but it failed again so it doesn't seem like a flake.

**Comment by soares-sergio** [comment]
> It seems there is an issue with `WaterHeaterMode.Attributes.SupportedModes` attribute:
> 
> ```
> [2024-11-07 17:54:28.881678][TEST][STDERR]INFO:chip.native.FP:Forgetting fabric 0x1
> [2024-11-07 17:54:28.882097][TEST][STDOUT]mobly.signals.TestFailure: Details=Error reading <class 'chip.clusters.Objects.WaterHeaterMode'>:<class 'chip.clusters.Objects.WaterHeaterMode.Attributes.SupportedModes'> = ValueDecodeFailure(TLVValue=None, Reason=InteractionModelError(<Status.UnsupportedEndpoint: 127>)), Extras=None
> [2024-11-07 17:54:28.882637][TEST][STDOUT][MatterTest] 11-07 17:54:28.873 INFO 		***** Test Failure : 
> [2024-11-07 17:54:28.882724][TEST][STDERR]INFO:chip.native.TS:Pending Last Known Good Time: 2023-10-14T01:16:48
> [2024-11-07 17:54:28.863686][APP ][STDOUT][25075:25075] CHIP:EM: Rxd Ack; Removing MessageCounter:265997561 from Retrans Table on exchange 3726r
> [2024-11-07 17:54:28.883327][TEST][STDOUT][MatterTest] 11-07 17:54:28.873 INFO Finished test in 8564.0ms
> [2024-11-07 17:54:28.883629][TEST][STDERR]INFO:chip.native.TS:Previous Last Known Good Time: 2023-10-14T01:16:48
> [2024-11-07 17:54:28.864435][APP ][STDOUT][25075:25075] CHIP:ATM: {
> [2024-11-07 17:54:28.883748][TEST][STDOUT][MatterTest] 11-07 17:54:28.873 ERROR 
> ```

Thanks, @lboue and @kiel-apple for double checking the REPL tests. I found the culprit!

`TC_WHM_2_1` passes when I run locally:
```
[2024-11-20 15:23:35.890526][TEST][STDOUT][MatterTest] 11-20 15:23:35.890 INFO Summary for test class TC_WHM_2_1: Error 0, Executed 1, Failed 0, Passed 1, Requested 1, Skipped 0
```
With this command:
```
rm /tmp/chip_* && scripts/run_in_python_env.sh out/python_env './scripts/tests/run_python_test.py --app ./out/linux-x64-energy-management-no-ble/chip-energy-management-app --app-args "--discriminator 1234 --featureSet 0x00 --application water-heater --trace-to json:log" --script src/python_testing/TC_WHM_2_1.py --script-args "--commissioning-method on-network --discriminator 1234 --passcode 20202021 --endpoint 2"'
```

So that pointed me to likely an issue with the CI scripts itself. I investigated a bit and found out the `--application` argument was incorrect in the test script itself. I fixed it in: https://github.com/project-chip/connectedhomeip/pull/36578 once that lands, I'll merge master on this PR and it should start passing the REPL tests.

**Comment by lboue** [comment]
> So that pointed me to likely an issue with the CI scripts itself. I investigated a bit and found out the `--application` argument was incorrect in the test script itself. I fixed it in: #36578 once that lands, I'll merge master on this PR and it should start passing the REPL tests.

@soares-sergio Thank you for your patches. The `Tests / REPL Tests - Linux` CI test is now working.

File: examples/energy-management-app/energy-management-common/common/src/EnergyManagementAppCommonMain.cpp
Comment (jamesharrow) [inline examples/energy-management-app/energy-management-common/common/src/EnergyManagementAppCommonMain.cpp:67]: I was (and may suggest that we still do) move this to a new EnergyReportingMain.cpp so that PowerTopology/EEM/EPM are moved into a re-usable component. That's a change I have in my 'refactoring' PR - but let's get this PR merged.

I think I had a DEMMain as well so that the DEM related code isn't necessarily in here.

Comment (soares-sergio) [reply examples/energy-management-app/energy-management-common/common/src/EnergyManagementAppCommonMain.cpp:67]: Agreed, that's a good idea. Let's do that in a follow up smaller PR.

File: examples/energy-management-app/energy-management-common/device-energy-management/src/DEMTestEventTriggers.cpp
Comment (PeterC1965) [inline examples/energy-management-app/energy-management-common/device-energy-management/src/DEMTestEventTriggers.cpp:37]: OOC why remove the static keyword?

Comment (soares-sergio) [reply examples/energy-management-app/energy-management-common/device-energy-management/src/DEMTestEventTriggers.cpp:37]: In this case, static is redundant, because these variables are in an anonymous namespace which limits the visibility of definitions to a single translation unit.

**Comment by PeterC1965** [comment]
This all looks good to me.

**Comment by lboue** [comment]
This branch has conflicts with the file [src/python_testing/TC_WHM_1_2.py](https://github.com/project-chip/connectedhomeip/blob/master/src/python_testing/TC_WHM_1_2.py)

**Comment by soares-sergio** [comment]
> This branch has conflicts with the file [src/python_testing/TC_WHM_1_2.py](https://github.com/project-chip/connectedhomeip/blob/master/src/python_testing/TC_WHM_1_2.py) (commit [926ab43](https://github.com/project-chip/connectedhomeip/commit/926ab430801b4158a8d75b1daf1b1adccf230639))

Just fixed it, thanks! 

And thanks for reviewing @PeterC1965 .
I'll get this landed.

**Pull request merged**
